### PR TITLE
Add `GetResource` gRPC service and endpoint.

### PR DESCRIPTION
**Pull request by jialez0**

### CODE REVIEW

File: protos/getresource.proto [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+package getresource;
+
+message getResourceRequest {
+    string KbcName = 1;
+    string KbsUri = 2;
+    string ResourceDescription = 3;
+}
+
+message getResourceResponse {
+    bytes resource = 1;
```
Comment (fitzthum) [inline protos/getresource.proto:12]: resource -> Resource?

File: src/kbc_runtime.rs [hunk 2]
```diff
@@ -54,6 +54,23 @@ impl KbcRuntime {
...
+        resource_description: String,
+    ) -> Result<Vec<u8>> {
+        if self.kbc_instance_map.get_mut(&kbc_name).is_none() {
+            self.instantiate_kbc(kbc_name.clone(), kbs_uri)?;
+        }
+        let kbc_instance = self
+            .kbc_instance_map
+            .get_mut(&kbc_name)
+            .ok_or(anyhow!("KBC runtime: The KBC instance does not existing!"))?;
+        let resource = kbc_instance.get_resource(resource_description)?;
```
Comment (fitzthum) [inline src/kbc_runtime.rs:70]: How can we handle KBCs that don't have a get_resource endpoint? I think we might want to add that endpoint as a stub to every existing KBC and create some mechanism for reporting that the feature isn't supported.

File: src/main.rs [hunk 3]
```diff
@@ -20,23 +20,42 @@ async fn main() -> Result<()> {
...
-                .long("grpc_sock")
+            Arg::with_name("KeyProvider gRPC socket addr")
+                .long("keyprovider_sock")
                 .takes_value(true)
-                .help("The socket address which the grpc service will listen to, for example: --grpc_sock 127.0.0.1:11223",
+                .help("This socket address which the KeyProvider gRPC service will listen to, for example: --keyprovider_sock 127.0.0.1:11223",
+                ),
+        )
+        .arg(
+            Arg::with_name("GetResource gRPC socket addr")
```
Comment (fitzthum) [inline src/main.rs:30]: Why create two different sockets for this? In practice won't `ocicrypt-rs` be the one using both? I guess it might be more robust to have a separate socket?

File: src/grpc/keyprovider/message.rs [hunk 4]
```diff
@@ -52,17 +52,20 @@ impl KeyProviderInput {
         Ok(())
     }
 
+    #[must_use]
```
Comment (fitzthum) [inline src/grpc/keyprovider/message.rs:55]: You might want to put this stuff in a separate PR. Should be trivial to merge.

File: protos/getresource.proto [hunk 5]
```diff
@@ -0,0 +1,17 @@
+syntax = "proto3";
+
+package getresource;
+
+message getResourceRequest {
```
Comment (fitzthum) [inline protos/getresource.proto:5]: I am not familiar with the convention here. Why is the first letter lowercase?

**Review by fitzthum — COMMENTED** [review]
Now that the the `image-rs` proposals have seemed to settle, let's try to get this merged. Sorry for the delay. A few thoughts to start with.

File: src/main.rs (hunk 3 above)
Comment (jialez0) [reply src/main.rs:30]: My considerations of using two separate sockets are as follows:

1. In fact, the caller of get resource is not ocicrypt RS, but image rs. ocicrypt RS does not support container image signature verification, but only image encryption and decryption. Image RS actually completes image signature verification. (this follows the implementation logic of ocicrypt and image in golang language version)

2. The protobuf of the key provider grpc service has been defined by ocicrypt. In order to maintain compatibility as much as possible, we'd better not change it.

File: src/kbc_runtime.rs (hunk 2 above)
Comment (jialez0) [reply src/kbc_runtime.rs:70]: In the trait definition of KBC interface, we can add the default implementation of KBC's get resource interface: https://github.com/confidential-containers/attestation-agent/pull/43/commits/4f3ae318b4fde38b637ac38bb41992987900177a 

For KBC without custom implementation of get resource interface, this default implementation will be used.

File: src/kbc_modules/mod.rs [hunk 6]
```diff
@@ -24,6 +24,9 @@ use std::sync::Arc;
 pub trait KbcInterface {
     fn check(&self) -> Result<KbcCheckInfo>;
     fn decrypt_payload(&mut self, annotation: &str) -> Result<Vec<u8>>;
+    fn get_resource(&self, _description: String) -> Result<Vec<u8>> {
+        Err(anyhow!("Get Resource API of this KBC is unimplement!"))
```
Comment (fitzthum) [inline src/kbc_modules/mod.rs:28]: nit: `unimplement` -> `not implemented`

**Review by fitzthum — APPROVED** [review]
LGTM. This PR seems fairly straightforward. I have a few minor comments. 

We'll need to adjust the Kata Agent slightly to set the additional socket. Let's remember to do that sometime.

The format of the resource description will require some thinking, but this PR doesn't determine that.

File: src/main.rs [hunk 7]
```diff
@@ -20,23 +20,42 @@ async fn main() -> Result<()> {
...
+        "GetResource gRPC service listening on: {:?}",
+        getresource_socket
+    );
 
-    grpc::start_service(socket).await?;
+    let keyprovider_server = grpc::keyprovider::start_service(keyprovider_socket);
+    let getresource_server = grpc::getresource::start_service(getresource_socket);
 
-    Ok(())
+    tokio::join!(keyprovider_server, getresource_server).0
```
Comment (fitzthum) [inline src/main.rs:60]: This is slightly weird. `.0` gives us the status of `keyprovider_server` only?

**Comment by liangzhou121** [comment]
>The format of the resource description will require some thinking, but this PR doesn't determine that.

Hi @fitzthum ,
We have created a new issue to discuss the definition of `resource description`. Please help to have a review also.
Issue: https://github.com/confidential-containers/guest-components/issues/243

File: protos/getresource.proto [hunk 8]
```diff
@@ -0,0 +1,17 @@
+syntax = "proto3";
+
+package getresource;
+
+message GetResourceRequest {
+    string KbcName = 1;
+    string KbsUri = 2;
```
Comment (jodh-intel) [inline protos/getresource.proto:7]: I'd consider documenting all message fields. For example, is this a real URL or a "fake" one? See https://github.com/confidential-containers/attestation-agent/issues/36.

File: protos/getresource.proto [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+    string ResourceDescription = 3;
+}
+
+message GetResourceResponse {
+    bytes Resource = 1;
```
Comment (jodh-intel) [inline protos/getresource.proto:12]: What exactly _is_ `Resource`?

File: src/grpc/getresource/mod.rs [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+pub mod get_resource {
+    tonic::include_proto!("getresource");
+}
+
+#[derive(Debug, Default)]
+pub struct GetResource {}
+
+#[tonic::async_trait]
+impl GetResourceService for GetResource {
```
Comment (jodh-intel) [inline src/grpc/getresource/mod.rs:25]: This looks odd to me. Wouldn't `Resource` be a better name for the `struct` which would then become:

```rust
impl GetResourceService for Resource
```

File: src/grpc/getresource/mod.rs [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+        &self,
+        request: Request<GetResourceRequest>,
+    ) -> Result<Response<GetResourceResponse>, Status> {
+        let request = request.into_inner();
+
+        let kbc_runtime_mutex_clone = Arc::clone(&kbc_runtime::KBC_RUNTIME);
+        let mut kbc_runtime = kbc_runtime_mutex_clone.lock().map_err(|e| {
+            error!("Get KBC runtime MUTEX failed: {}", e);
+            Status::internal(format!(
+                "[ERROR:{}] Get KBC runtime failed: {}",
```
Comment (jodh-intel) [inline src/grpc/getresource/mod.rs:36]: Not required for this PR, but I think we need a macro to avoid the constant use of the `[ERROR:` prefix.

**Review by jodh-intel — COMMENTED** [review]
Thanks @jialez0. A few comments...

File: src/grpc/keyprovider/message.rs (hunk 4 above)
Comment (jialez0) [reply src/grpc/keyprovider/message.rs]: Done.

**Comment by fitzthum** [comment]
Ok, I guess we are waiting for `image-rs` support to get merged?

**Comment by jialez0** [comment]
\tests

**Comment by jialez0** [comment]
I think this PR can get merged first and don't need to wait for image-rs, so that we can start adding KBC support for get resource :)
